src/sparse/csmat.rs

```diff
@@ -1139,6 +1139,30 @@ where
         })
     }
 
+    /// Get the diagonal of a sparse matrix
+    pub fn diag(&self) -> CsVecI<N, I>
+    where
+        N: Clone,
+    {
+        let shape = self.shape();
+        let smallest_dim: usize = cmp::min(shape.0, shape.1);
+        let mut index_vec = Vec::new();
+        let mut data_vec = Vec::new();
+        let mut optional_index: Option<NnzIndex>;
```

Comment:
Looks like `optional_index` does not need to be mutable, and could be declared inside the loop.

Reply:
It's actually exactly the same performance. You are right that there are cases where it would be more performant to have a mutable variable declared outside of the loop, but here we have a Plain Old Data type, so whichever way it is declared the compiler will simply reserve the required stack storage (allocation of stack storage is per-function).

An example of pattern where there would be a difference is with a type that involves heap allocation such as `Vec`:

```rust
let mut v = Vec::with_capacity(10);
for i in 0..10 {
    v.clear();
    v.extend(0..i);
   println!("sum: {}", v.iter().sum());
}
```

is more performant than

```rust
for i in 0..10 {
    let mut v = Vec::new();
    v.extend(0..i);
   println!("sum: {}", v.iter().sum());
}
```

because the first version is able to re-use the same heap storage over and over again, while the second one reallocates at each step of the loop.

Reply:
Ah, I see, thanks for the explenation